async-commit-the-accelerator-for-transaction-commit-in-tidb-5.0.md

```diff
@@ -0,0 +1,326 @@
+
+<table>
+  <tr>
+   <td>T1: Begin(Start TS = 1)
+   </td>
+   <td>
+   </td>
+  </tr>
+  <tr>
+   <td>T1: Prewrite(x)
+   </td>
+   <td>T2: Begin(Start TS = 5)
+   </td>
+  </tr>
+  <tr>
+   <td>
+   </td>
+   <td>T2: Read(y) => Max TS = 5
+   </td>
+  </tr>
+  <tr>
+   <td>T1: Prewrite(y) => Min Commit TS = 6
+   </td>
+   <td>
+   </td>
+  </tr>
+  <tr>
+   <td>
+   </td>
+   <td>T2: Read(y)
+   </td>
+  </tr>
+</table>
+
+### Guarantee linearizability
+
+Linearizability includes two aspects:
+
+* Real-time
+* Sequential consistency
+
+The aspect of "real-time" requires that transaction changes must be able to be read by a new transaction immediately after the previous transaction successfully commits. The snapshot timestamp of the new transaction is obtained from the TSO on PD, which requires the Commit TS not be too large. The value of Commit TS can only be up to the maximum timestamp allocated by the TSO plus one.
+
+As mentioned in the snapshot isolation section, one possible value for Min Commit TS is Max TS + 1. The timestamps used to update Max TS are from the TSO, so Max TS + 1 is definitely less than or equal to the minimum unallocated timestamp on the TSO. Thus, the requirement of "real-time" is satisfied.
+
+Sequential consistency means that the logical order of transactions does not violate the physical order of precedence. For example, let's assume that there are two transactions, T1 and T2. If T1 is committed before T2, T1's Commit TS should be smaller than or equal to that of T2.[^3]
+
+To guarantee sequential consistency, TiDB gets a timestamp from TSO on PD as the minimum constraint of Min Commit TS before the prewriting process. Because the real-time property is guaranteed, the timestamp obtained by T2 before the prewriting process must be greater than or equal to the Commit TS of T1. Therefore, we can guarantee that no causal reverse happens, and sequential consistency is guaranteed.
+
+In summary, the Min Commit TS of each key takes the maximum value between the Max TS + 1 that TiDB gets when prewriting, and the timestamp obtained from the PD before the prewriting process. The Commit TS of a transaction is the maximum Min Commit TS of all keys. They together guarantee both snapshot isolation and linearizability for Async Commit transactions.
+
+### One-phase commit (1PC)
+
+If a transaction only updates a non-index column of a record, or inserts a record without a secondary index, it only involves a single region. In this scenario, is it possible to complete the transaction commit in only one phase without using the distributed transaction protocol? Well, this is certainly possible, but the difficulty lies in how to determine the Commit TS for a one-phase commit transaction.
```

Comment:
```suggestion
If a transaction only updates a non-index column of a record, or inserts a record without a secondary index, it only involves a single [Region](https://docs.pingcap.com/tidb/stable/glossary#regionpeerraft-group). In this scenario, is it possible to complete the transaction commit in only one phase without using the distributed transaction protocol? Well, this is certainly possible, but the difficulty lies in how to determine the Commit TS for a one-phase commit transaction.
```

Reply:
"Region" is a technical term for TiDB. Please update all occurrences. Thanks.

Reply:
Updated in https://github.com/pingcap/blog/pull/581/commits/54d3197c266d03f140d0abd3c62b1c4261eda1a3.